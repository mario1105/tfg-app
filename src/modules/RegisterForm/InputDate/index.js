import React, { useEffect } from 'react';
import {
  Input, FormControl, InputLabel, makeStyles,
} from '@material-ui/core';
import PropTypes from 'prop-types';
import Typography from 'components/Typography';

const useStyles = makeStyles(() => ({
  formControlContainer: {
    display: 'flex',
    marginBottom: '.5em'
  },
}));

const setFocus = (inputId) => {
  setTimeout(() => document.getElementById(inputId).focus(), 1);
};

const InputDate = ({
  date, errors, handleFieldsChange, handleFieldsErrors, resetErrors, ...props
}) => {
  const { day, month, year } = date;

  useEffect(
    () => {
      if (day.length === 2) {
        if (month.length !== 2) setFocus('month-input');
        else if (month.length === 2 && year.length !== 4) setFocus('year-input');
      } else if (day.length === 0 && month.length === 2) setFocus('year-input');
    },
    [day, month],
  );

  const classes = useStyles();

  return (
    <div>
      <div className={classes.formControlContainer}>
        <FormControl data-test-id="date-field-day-form-control" error={errors.day} {...props}>
          <InputLabel>DD</InputLabel>
          <Input
            data-test-id="date-field-day-input"
            id={'day-input'}
            type={'number'}
            className={'pureNumberInput'}
            autoComplete="off"
            value={day}
            onChange={handleFieldsChange('day')}
            onBlur={() => handleFieldsErrors('day')}
            onFocus={() => resetErrors('day')}
            onInput={(e) => {
              e.target.value = e.target.value.slice(0, 2);
            }}
          />
        </FormControl>
        <FormControl data-test-id="date-field-month-form-control" error={errors.month} {...props}>
          <InputLabel>MM</InputLabel>
          <Input
            data-test-id="date-field-month-input"
            id={'month-input'}
            type={'number'}
            className={'pureNumberInput'}
            autoComplete="off"
            value={month}
            onChange={handleFieldsChange('month')}
            onBlur={() => handleFieldsErrors('month')}
            onFocus={() => resetErrors('month')}
            onInput={(e) => {
              e.target.value = e.target.value.slice(0, 2);
            }}
          />
        </FormControl>
        <FormControl data-test-id="date-field-year-form-control" error={errors.year} {...props}>
          <InputLabel>YYYY</InputLabel>
          <Input
            data-test-id="date-field-year-input"
            id={'year-input'}
            type={'number'}
            className={'pureNumberInput'}
            autoComplete="off"
            value={year}
            onChange={handleFieldsChange('year')}
            onBlur={() => handleFieldsErrors('year')}
            onFocus={() => resetErrors('year')}
            onInput={(e) => {
              e.target.value = e.target.value.slice(0, 4);
            }}
          />
        </FormControl>
      </div>
      <Typography variant="subtitle2" gutterBottom>
        Date of Birth
      </Typography>
    </div>

  );
};

InputDate.propTypes = {
  date: PropTypes.shape({
    day: PropTypes.string,
    month: PropTypes.string,
    year: PropTypes.string
  }),
  errors: PropTypes.shape({
    day: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    month: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    year: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  }),
  handleFieldsChange: PropTypes.func,
  handleFieldsErrors: PropTypes.func,
  resetErrors: PropTypes.func,
};

export default InputDate;
