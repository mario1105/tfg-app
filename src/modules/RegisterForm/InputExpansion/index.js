import React, { useState } from 'react';
import PropTypes from 'prop-types';
import InputExpansion from './InputExpansion';


const InputExpansionContainer = ({ content, children, ...props }) => {
  const [expanded, setExpanded] = useState(false);
  const handleChange = (event = { target: { className: {} } }) => {
    if (event.target.tagName === 'svg' || event.target.tagName === 'path' || event.target.className.includes('MuiButtonBase-root')) {
      setExpanded(!expanded);
    }
  };
  return (
    <InputExpansion content={content} expanded={expanded} handleChange={handleChange} {...props}>
      {children}
    </InputExpansion>
  );
};

InputExpansionContainer.propTypes = {
  content: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  children: PropTypes.node.isRequired
};

export default InputExpansionContainer;
