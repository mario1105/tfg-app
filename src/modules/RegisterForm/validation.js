const validation = {
  email: (email) => !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email) && 'Please enter a valid email',
  firstName: (firstName) => firstName.length < 2 && 'First name must be at least 2 characters long',
  lastName: (lastName) => lastName.length < 2 && 'Last name must be at least 2 characters long',
  mobilePhone: (mobilePhone) => (
    (!/^\+[1-9]\d{1,14}$/.test(mobilePhone)
    || !((mobilePhone.startsWith('+447') && mobilePhone.length === 13)
      || (mobilePhone.startsWith('+4407') && mobilePhone.length === 14)))
    && 'Please enter a valid UK phone number.'),
  day: (day) => day.length !== 2 || day > 31 || day === '00',
  month: (month) => month.length !== 2 || month > 12 || month === '00',
  year: (year) => year.length !== 4 || year < 1900 || year === '0000'
};

export default validation;
