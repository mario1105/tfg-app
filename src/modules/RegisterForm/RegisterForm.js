import React from 'react';
import * as R from 'ramda';
import {
  Grid, makeStyles, Input, FormControl, InputLabel, FormHelperText, Typography, Button
} from '@material-ui/core';
import InputExpansion from './InputExpansion';
import InputDate from './InputDate';
import ReturnButton from "../../components/ReturnButton"

const useStyles = makeStyles((theme) => ({
  card: {
    backgroundColor: theme.app.background['300'],
    borderRadius: '20px',
    maxWidth: '25em',
    margin: 'auto',
  },
  form: {
    display: 'flex',
    flexDirection: 'column',
    margin: '1em'
  },
  formControl: {
    flex: '1',
    paddingBottom: '1em'
  },
}));

const RegisterForm = ({
                        user, fields, errors, loading, confirmationView, requestError, handleFieldsChange, handleFieldsErrors, resetErrors, submitRegisterForm
}) => {
  const validationError = errors.email || errors.firstName || errors.lastName || errors.mobilePhone || errors.day || errors.month || errors.year;
  const emptyField = R.any((value) => value.length === 0)(R.values(fields));

  const classes = useStyles();
  return (
    <Grid className={classes.registerFormBox} container style={{ marginTop: '2em' }} alignItems="center" direction="column">
      <Grid item className={classes.card} style={{ marginBottom: '1em' }}>
          {confirmationView
            ? (
              <Grid data-test-id="registration-completed-view" style={{ margin: '1em', display: 'flex', flexDirection: 'column' }}>
                <Typography style={{ color: 'white' }}>Registration successfully completed</Typography>
                <ReturnButton previousRoute="dashboard" user={user} style={{ marginTop: '2em' }} />
              </Grid>
            )
            : (
              <form data-test-id="registration-form" noValidate className={classes.form} onSubmit={submitRegisterForm}>
                <FormControl error={!!errors.email} className={classes.formControl} style={{ paddingRight: '38px' }}>
                  <InputLabel htmlFor="my-input">Email address</InputLabel>
                  <Input
                    data-test-id="registration-form-email-input"
                    autoComplete="off"
                    value={fields.email}
                    onChange={handleFieldsChange('email')}
                    onBlur={() => handleFieldsErrors('email')}
                    onFocus={() => resetErrors('email')}
                  />
                  {!!errors.email && <FormHelperText data-test-id="registration-form-email-error">{errors.email}</FormHelperText>}
                </FormControl>

                <FormControl error={!!errors.firstName} className={classes.formControl} style={{ paddingRight: '38px' }}>
                  <InputLabel htmlFor="my-input">First name</InputLabel>
                  <Input
                    data-test-id="registration-form-first-name-input"
                    autoComplete="off"
                    value={fields.firstName}
                    onChange={handleFieldsChange('firstName')}
                    onBlur={() => handleFieldsErrors('firstName')}
                    onFocus={() => resetErrors('firstName')}
                  />
                  {!!errors.firstName && <FormHelperText data-test-id="registration-form-first-name-error">{errors.firstName}</FormHelperText>}
                </FormControl>

                <FormControl error={!!errors.lastName} className={classes.formControl} style={{ paddingRight: '38px' }}>
                  <InputLabel htmlFor="my-input">Last name</InputLabel>
                  <Input
                    data-test-id="registration-form-last-name-input"
                    autoComplete="off"
                    value={fields.lastName}
                    onChange={handleFieldsChange('lastName')}
                    onBlur={() => handleFieldsErrors('lastName')}
                    onFocus={() => resetErrors('lastName')}
                  />
                  {!!errors.lastName && <FormHelperText data-test-id="registration-form-last-name-error">{errors.lastName}</FormHelperText>}
                </FormControl>

                <InputExpansion
                  data-test-id="registration-form-mobile-phone-expansion-field"
                  content={(
                    <Typography variant="subtitle2" component={'span'}>
                      We require a Spanish phone number for additional contact details
                    </Typography>
                  )}
                >
                  <FormControl
                    data-test-id="registration-form-mobile-phone-field"
                    error={!!errors.mobilePhone}
                    className={classes.formControl}
                  >
                    <InputLabel htmlFor="my-input">Mobile phone</InputLabel>
                    <Input
                      data-test-id="registration-form-mobile-phone-input"
                      id={'mobile-phone-input'}
                      autoComplete="off"
                      value={fields.mobilePhone}
                      onChange={handleFieldsChange('mobilePhone')}
                      onBlur={() => handleFieldsErrors('mobilePhone')}
                      onFocus={() => resetErrors('mobilePhone')}
                    />
                    {!!errors.mobilePhone && <FormHelperText data-test-id="registration-form-mobile-phone-error">{errors.mobilePhone}</FormHelperText>}
                  </FormControl>
                </InputExpansion>
                <InputDate
                  date={{ day: fields.day, month: fields.month, year: fields.year }}
                  errors={{ day: errors.day, month: errors.month, year: errors.year }}
                  handleFieldsChange={handleFieldsChange}
                  handleFieldsErrors={handleFieldsErrors}
                  resetErrors={resetErrors}
                  style={{ paddingRight: '38px' }}
                />

                <FormControl error={!!errors.salary} className={classes.formControl} style={{ paddingRight: '38px' }}>
                  <InputLabel htmlFor="my-input">Salary</InputLabel>
                  <Input
                      data-test-id="registration-form-salary-input"
                      autoComplete="off"
                      value={fields.salary}
                      onChange={handleFieldsChange('salary')}
                      onBlur={() => handleFieldsErrors('salary')}
                      onFocus={() => resetErrors('salary')}
                  />
                  {!!errors.salary && <FormHelperText data-test-id="registration-form-salary-error">{errors.salary}</FormHelperText>}
                </FormControl>
                <Button
                  data-test-id="registration-form-register-button"
                  disabled={!!(loading || validationError || emptyField)}
                  style={{ marginTop: '3em' }}
                  color="primary"
                  variant="outlined"
                  type="submit"
                  fullWidth
                  onClick={submitRegisterForm}
                >Register
                </Button>
                {requestError
                && (
                  <Typography
                    data-test-id="registration-form-submit-error"
                    style={{ display: 'flex', alignSelf: 'center' }}
                    color="error"
                  >{requestError}
                  </Typography>
                )}
                <ReturnButton previousRoute="dashboard" user={user} style={{ marginTop: '2em' }} />
              </form>
            )
          }
      </Grid>
    </Grid>
  )
}

export default RegisterForm;
