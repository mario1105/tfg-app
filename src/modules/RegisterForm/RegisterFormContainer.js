import React, { useState } from 'react';
import * as R from 'ramda';
import request from 'utils/request';
import PropTypes from 'prop-types';
import RegisterForm from './RegisterForm';
import validation from './validation';
import getOptions from './service/getOptions';
import apiEndpointBase from './service/apiEndpointBase';

const RegisterFormContainer = ({ userEmail }) => {
  const [fields, setFields] = useState({
    email: '',
    firstName: '',
    lastName: '',
    mobilePhone: '+44',
    day: '',
    month: '',
    year: ''
  });

  const [errors, setErrors] = useState({
    email: false,
    firstName: false,
    lastName: false,
    mobilePhone: false,
    day: false,
    month: false,
    year: false
  });

  const [loading, setLoading] = useState(false);
  const [confirmationView, setConfirmationView] = useState(false);
  const [requestError, setRequestError] = useState(false);


  const handleFieldsChange = (prop) => (event) => {
    if (prop === 'mobilePhone' && !event.target.value.startsWith('+44')) {
      setFields({ ...fields, mobilePhone: '+44' });
    } else {
      setFields({ ...fields, [prop]: event.target.value });
    }
  };

  const handleFieldsErrors = (prop) => {
    setErrors({ ...errors, [prop]: validation[prop](fields[prop]) });
  };

  const resetErrors = (prop) => {
    setErrors({ ...errors, [prop]: false });
  };

  const submitRegisterForm = async (e) => {
    e.preventDefault();
    setRequestError(false);

    setLoading(true);
    await request(`${apiEndpointBase}/register`, getOptions(fields))
      .then(() => {
        setLoading(false);
        setConfirmationView(true);
      })
      .catch((error) => {
        setLoading(false);
        if (error.response) {
          error.response.json().then((data) => {
            const validations = data.errors.map(({ meta }) => meta.validation);
            const emailAlreadyExists = R.flatten(validations).find(({ failed }) => failed.Okta === 'An object with this field already exists in the current organization');
            // eslint-disable-next-line no-unused-expressions
            emailAlreadyExists ? setRequestError('This email address already exists') : setRequestError('There was an error, please try again');
          });
        } else setRequestError('There was an error, please try again');
      });
  };

  return (
    <RegisterForm
      userEmail={userEmail}
      fields={fields}
      errors={errors}
      loading={loading}
      confirmationView={confirmationView}
      requestError={requestError}
      handleFieldsChange={handleFieldsChange}
      handleFieldsErrors={handleFieldsErrors}
      resetErrors={resetErrors}
      submitRegisterForm={submitRegisterForm}
    />
  );
};

RegisterFormContainer.propTypes = {
  userEmail: PropTypes.string,
};
export default RegisterFormContainer;
