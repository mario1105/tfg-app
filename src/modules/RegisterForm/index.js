import React, { useState } from 'react'
import * as R from 'ramda'
import RegisterForm from './RegisterForm'
import validation from './validation'
import employeesService from "../../services/employeesService"

const RegisterFormContainer = ({location}) => {
    const user = location.state

    const [fields, setFields] = useState({
        email: '',
        firstName: '',
        lastName: '',
        mobilePhone: '',
        day: '',
        month: '',
        year: '',
        salary: ''
    })

    const [errors, setErrors] = useState({
        email: false,
        firstName: false,
        lastName: false,
        mobilePhone: false,
        day: false,
        month: false,
        year: false,
        salary: false
    })

    const [loading, setLoading] = useState(false)
    const [confirmationView, setConfirmationView] = useState(false)

    const handleFieldsChange = (prop) => (event) => setFields({ ...fields, [prop]: event.target.value })

    const handleFieldsErrors = (prop) => setErrors({ ...errors, [prop]: validation[prop](fields[prop]) })

    const resetErrors = (prop) => setErrors({ ...errors, [prop]: false })

    const submitRegisterForm = async (e) => {
        e.preventDefault()

        setLoading(true)
        await employeesService.registerEmployee({
            firstName: fields.firstName,
            lastName: fields.lastName,
            email: fields.email,
            dateOfBirth: R.join('/', [fields.day, fields.month, fields.year]),
            mobilePhone: fields.mobilePhone,
            salary: fields.salary
        })
        setLoading(false)
        setConfirmationView(true)
    }

    return (
        <RegisterForm
            user={user}
            fields={fields}
            errors={errors}
            loading={loading}
            confirmationView={confirmationView}
            handleFieldsChange={handleFieldsChange}
            handleFieldsErrors={handleFieldsErrors}
            resetErrors={resetErrors}
            submitRegisterForm={submitRegisterForm}
        />
    )
}

export default RegisterFormContainer
