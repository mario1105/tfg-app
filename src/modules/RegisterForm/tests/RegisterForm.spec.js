import React from 'react'
import { shallow } from 'enzyme'
import RegisterForm from '../RegisterForm'

const defaultProps = {
  user: {},
  fields: {
    email: 'mario.jimenez@theshopworks.com',
    firstName: 'Mario',
    lastName: 'Jimenez',
    mobilePhone: '+444444444444',
    day: '11',
    month: '11',
    year: '1999',
    salary: '50000'
  },
  errors: {
    email: false,
    firstName: false,
    lastName: false,
    mobilePhone: false,
    day: false,
    month: false,
    year: false,
    salary: false,
  },
  loading: false,
  confirmationView: false,
  requestError: false,
  handleFieldsChange: () => null,
  handleFieldsErrors: () => null,
  submitRegisterForm: () => null,
  resetErrors: () => null
}

const renderComponent = (newProps) => {
  const props = {
    ...defaultProps,
    ...newProps,
  }

  return (shallow(<RegisterForm {...props} />))
}

let wrapper
describe('RegisterForm', () => {
  describe('when there are no errors', () => {
    beforeEach(() => {
      wrapper = renderComponent()
    })

    it('renders as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })
    it('sets the input values with the correct prop', () => {
      expect(wrapper.find('[data-test-id~="registration-form-email-input"]').prop('value')).toBe(defaultProps.fields.email)
      expect(wrapper.find('[data-test-id~="registration-form-first-name-input"]').prop('value')).toBe(defaultProps.fields.firstName)
      expect(wrapper.find('[data-test-id~="registration-form-last-name-input"]').prop('value')).toBe(defaultProps.fields.lastName)
      expect(wrapper.find('[data-test-id~="registration-form-mobile-phone-input"]').prop('value')).toBe(defaultProps.fields.mobilePhone)
      expect(wrapper.find('[data-test-id~="registration-form-salary-input"]').prop('value')).toBe(defaultProps.fields.salary)
    })
    it('passes the correct props to InputDate', () => {
      expect(wrapper.find('InputDate').prop('date')).toEqual({ day: defaultProps.fields.day, month: defaultProps.fields.month, year: defaultProps.fields.year })
      expect(wrapper.find('InputDate').prop('errors')).toEqual({ day: defaultProps.errors.day, month: defaultProps.errors.month, year: defaultProps.errors.year })
      expect(wrapper.find('InputDate').prop('handleFieldsChange')).toBe(defaultProps.handleFieldsChange)
      expect(wrapper.find('InputDate').prop('handleFieldsErrors')).toBe(defaultProps.handleFieldsErrors)
      expect(wrapper.find('InputDate').prop('resetErrors')).toBe(defaultProps.resetErrors)
    })
  })
  describe('when there are errors', () => {
    it('renders as expected', () => {
      expect(renderComponent({
        errors: {
          email: 'Email error',
          firstName: 'First name error',
          lastName: 'Last name error',
          mobilePhone: 'Mobile phone error',
          day: 'Day error',
          month: 'Month error',
          year: 'Year error',
          salary: 'Salary error'
        },
      })).toMatchSnapshot()
    })
  })
})
