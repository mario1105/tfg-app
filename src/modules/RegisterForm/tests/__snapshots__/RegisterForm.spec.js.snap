// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm when there are errors renders as expected 1`] = `ShallowWrapper {}`;

exports[`RegisterForm when there are no errors renders as expected 1`] = `ShallowWrapper {}`;
